src/core/radio/trel_link.cpp

```diff
@@ -373,8 +376,42 @@ void Link::ProcessReceivedPacket(Packet &aPacket)
 
     Get<Mac::Mac>().HandleReceivedFrame(&mRxFrame, kErrorNone);
 
+    // As the received frame is processed by the MAC or MLE layers,
+    // `CheckPeerAddrOnRxSuccess()` may be called with different modes,
+    // depending on whether the frame passes receive security checks
+    // at either the MAC or MLE layers, allowing or disallowing peer
+    // socket address to be updated from received TREL packet info.
+
 exit:
-    return;
+    mRxPacketPeer = nullptr;
+}
+
+void Link::CheckPeerAddrOnRxSuccess(PeerSockAddrUpdateMode aMode)
+{
+    Ip6::SockAddr prevSockAddr;
+
+    VerifyOrExit(mState != kStateDisabled);
+
+    VerifyOrExit(mRxPacketPeer != nullptr);
+
+    prevSockAddr = mRxPacketPeer->GetSockAddr();
+    VerifyOrExit(prevSockAddr != mRxPacketSenderAddr);
+
+    LogNote("Peer %s rx sock-addr differs from previously discovered by platform",
+            mRxPacketPeer->GetExtAddress().ToString().AsCString());
+    LogNote("    Rcvd sock-addr:%s", mRxPacketSenderAddr.ToString().AsCString());
+    LogNote("    Prev sock-addr:%s", prevSockAddr.ToString().AsCString());
+
+    if (aMode == kAllowPeerSockAddrUpdate)
+    {
+        LogNote("Updating the peer sock-addr to the newly received");
+        mRxPacketPeer->SetSockAddr(mRxPacketSenderAddr);
```

Comment:
can we add a hysteresis here to change the peer address based on consecutive received count?  
Wondering if there is a possibility to flip-flop addresses especially if the source address selection on the sender can prefer different address intermittently for some packets?

Reply:
I don't think this would be very useful, as I see no harm in changing the peer address. If we can receive from a TREL peer on an IPv6 address, we should be able to use the same address to send packets to it. Even if a TREL peer has multiple addresses and flip-flops between them, it should be reachable on any of its addresses, so it is okay for us to remember the last one.

Reply:
thanks for the clarification!

Reply:
If one TREL peer flip-flops between different source addresses, this would still generate a large log volume (NOTE level). Which would be generally unwanted. Though, not sure how likely this scenario is.